Sources/NIO/ByteBuffer-aux.swift

```diff
@@ -152,6 +155,51 @@ extension ByteBuffer {
         }
     }
 
+    // MARK: Substring APIs
+    /// Write `substring` into this `ByteBuffer` using UTF-8 encoding, moving the writer index forward appropriately.
+    ///
+    /// - parameters:
+    ///     - substring: The substring to write.
+    /// - returns: The number of bytes written.
+    @discardableResult
+    public mutating func writeSubstring(_ substring: Substring) -> Int {
+        let written = self.setSubstring(substring, at: self.writerIndex)
+        self._moveWriterIndex(forwardBy: written)
+        return written
+    }
+    
+    @inline(never)
+    @usableFromInline
+    mutating func _setSubstringSlowpath(_ substring: Substring, at index: Int) -> Int {
+        // slow path, let's try to force the substring to be native
+        if let written = (substring + "").utf8.withContiguousStorageIfAvailable({ utf8Bytes in
```

Comment:
@weissi You're the relevant expert here: is this a good idea on a `Substring`? It'll force an allocation and copy, so maybe we want to avoid this?

Reply:
unfortunately, this won't work at all on `Substring` (which is a bug). See this:

```swift
import Foundation
let nsString = NSString(data: Data([0x42, 0, 0x42, 0]), encoding: String.Encoding.utf16.rawValue)!

print("hello"                    .utf8.withContiguousStorageIfAvailable({ _ in "string direct ok" }) ?? "string direct NOT okay")
print((nsString as String)       .utf8.withContiguousStorageIfAvailable({ _ in "nsstring direct ok" }) ?? "nsstring direct NOT okay")
print(((nsString as String) + "").utf8.withContiguousStorageIfAvailable({ _ in "nsstring + '' direct ok" }) ?? "nsstring + '' direct NOT okay")
print("hello"[...]               .utf8.withContiguousStorageIfAvailable({ _ in "substring direct ok" }) ?? "substring direct NOT okay")
print(("hello"[...] + "")        .utf8.withContiguousStorageIfAvailable({ _ in "substring + '' direct ok" }) ?? "substring + '' direct NOT okay")
print(String("hello"[...])       .utf8.withContiguousStorageIfAvailable({ _ in "substring through string direct ok" }) ?? "substring through string direct NOT okay")
```

outputs

```
string direct ok
nsstring direct NOT okay
nsstring + '' direct ok
substring direct NOT okay
substring + '' direct NOT okay
substring through string direct ok
```

so I think the only way we can make sub strings behave is to always do `String(subString)` and then go through `writeString` for now. We should also file a bugs.swift.org about this.

Reply:
There’s already a bug: this is covered under the fact that substrings don’t have a `withContiguousStorageIfAvailable` implementation. But they _will_, and it’d be nice to not have to rewrite this when they do. 

The question is, is it acceptable to allocate a new `String` to get this fast path?

Reply:
I guess it’s interesting that `Substring + Substring` returns `Substring`.

Reply:
> The question is, is it acceptable to allocate a new `String` to get this fast path?

good question. Really, it depends on the length of the string. But I'd say we _should_ have a `writeSubstring` and for now we implement it as `String(sub)`. We can always improve later without breaking the API, right?

Reply:
That would work for me

Reply:
Sounds good. I'll make that change then.

Reply:
thank you!